### ISSUE TITLE
Information gets lost when saving and loading a model

### ISSUE THREAD

**Issue by hhoefener** [tensorflow/tensorflow#57917]
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**:
-   **TensorFlow installed from (source or binary)**: pip
-   **TensorFlow version (use command below)**: v2.7.0-rc1-69-gc256c071bb2 2.7.0
-   **Python version**: Python 3.8.10
-   **Bazel version (if compiling from source)**:
-   **GCC/Compiler version (if compiling from source)**:
-   **CUDA/cuDNN version**:
-   **GPU model and memory**:
-   **Exact command to reproduce**:

### Describe the problem
I create a keras model, train it, then save it. Documentation says that I can load the model and resume training right where I left off. That is not the case for me. My training loss always increases sharply on resuming training. Therefore I suspect that somewhere information about some state is lost.

### Source code / logs
I am training a model for few epochs, then save it. Now I expect training the same model another epochs gives the exact same results as loading the saved model and training another epoch there. I added some RNG seeding and other stuff to make training deterministic.
Here is a modified version of the keras MNIST example code that reproduces the behavior:

```python
#try making training deterministic

import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
seed_value= 123
os.environ['PYTHONHASHSEED']=str(seed_value)
import random
import numpy as np
import tensorflow as tf
tf.config.threading.set_intra_op_parallelism_threads(1)
tf.config.threading.set_inter_op_parallelism_threads(1)

def seed():
    random.seed(seed_value)
    np.random.seed(seed_value)
    tf.random.set_seed(seed_value)
seed()

# taken from MNIST example code (https://github.com/keras-team/keras-io/blob/master/examples/vision/mnist_convnet.py)

from tensorflow import keras
from tensorflow.keras import layers

"""
## Prepare the data
"""

# Model / data parameters
num_classes = 10
input_shape = (28, 28, 1)

# Load the data and split it between train and test sets
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

# Scale images to the [0, 1] range
x_train = x_train.astype("float32") / 255
x_test = x_test.astype("float32") / 255
# Make sure images have shape (28, 28, 1)
x_train = np.expand_dims(x_train, -1)
x_test = np.expand_dims(x_test, -1)
print("x_train shape:", x_train.shape)
print(x_train.shape[0], "train samples")
print(x_test.shape[0], "test samples")


# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

"""
## Build the model
"""

model = keras.Sequential(
    [
        keras.Input(shape=input_shape),
        layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Flatten(),
        layers.Dropout(0.5),
        layers.Dense(num_classes, activation="softmax"),
    ]
)

model.summary()

"""
## Train the model
"""

batch_size = 128

model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

model.fit(x_train, y_train, batch_size=batch_size, epochs=5)
model.save('model')
print('these two runs should be identical')
seed()
model.fit(x_train, y_train, batch_size=batch_size, epochs=1)
reloaded_model = keras.models.load_model('model')
seed()
reloaded_model.fit(x_train, y_train, batch_size=batch_size, epochs=1)
```
Which outputs on my machine:
```
...
these two runs should be identical
469/469 [==============================] - 33s 71ms/step - loss: 0.0509 - accuracy: 0.9842
469/469 [==============================] - 33s 69ms/step - loss: 0.0522 - accuracy: 0.9837
```

**Comment by hhoefener** [comment]
I am getting the exact same result (loss of 0.0522) if I save the model without optimizer and recompile after loading
```python
model.save('model', include_optimizer=False)
print('these two runs should be identical')
seed()
model.fit(x_train, y_train, batch_size=batch_size, epochs=1)
reloaded_model = keras.models.load_model('model')
reloaded_model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
seed()
reloaded_model.fit(x_train, y_train, batch_size=batch_size, epochs=1)
```

**Comment by tilakrayal** [comment]
@hhoefener,
The fact that an input pipeline with no unseeded random operations (and sources of external state such as shuffle or cache) produces the same order every time it is executed is an important invariant and the expected behavior (and changing it is not an option for backwards compatibility reasons).

**Comment by tilakrayal** [comment]
**tf.random.set_seed** sets the global random seed. Operations that rely on a random seed actually derive it from two seeds: the global and operation-level seeds. This sets the global seed.

1. If neither the global seed nor the operation seed is set: A randomly picked seed is used for this op.

2.  If the global seed is set, but the operation seed is not: The system deterministically picks an operation seed in conjunction with the global seed so that it gets a unique random sequence. Within the same version of tensorflow and user code, this sequence is deterministic. However across different versions, this sequence might change. If the code depends on particular seeds to work, specify both global and operation-level seeds explicitly.

3. If the operation seed is set, but the global seed is not set: A default global seed and the specified operation seed are used to determine the random sequence.

4. If both the global and the operation seed are set: Both seeds are used in conjunction to determine the random sequence.

**Comment by hhoefener** [comment]
@tilakrayal thank you for your response. The randomness and the seeding were not the issue here, I just added seeding to get reproducible runs to be able to compare the model created in the code with the saved-and-loaded model.
However, changing the save and load to h5 files, the issue is gone. So it seems to be a bug in the SavedModel format. Closing this issue in favor of this already existing one over at keras: https://github.com/keras-team/keras/issues/15512
